Expand collapse nodes

jweinstein92 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Nodes are one level deep
Issue Number: N/A


## What is the new behavior?
Allow nested/collapsible nodes by providing a parent id onto a Node. Clicking onto a node with children will dynamically generate edges between the parent and child node. Nodes can have multiple nested children

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

https://user-images.githubusercontent.com/6655992/186534474-c05af51f-c72d-4903-a201-1a6207ab6c9f.mov

amcdnl commented on src/GraphScene.tsx:
```diff
@@ -43,6 +45,11 @@ export interface GraphSceneProps {
+  /**
+   * List of node ids that are expanded.
+   */
+  expandedParents?: string[];
```
Lets invert this logic to:

```
collapsedNodeIds?: string[]
```

so that everything is expanded by default unless you specify otherwise and that should remove the need for the other `expanded` flag too.

amcdnl commented on src/GraphScene.tsx:
```diff
@@ -223,6 +256,21 @@ export const GraphScene: FC<GraphSceneProps & { ref?: Ref<GraphSceneRef> }> =
       );
 
+      useEffect(() => {
+        if (expandedParents.length > 0) {
```
This should probaly live in the useGraph hook

amcdnl commented on src/symbols/Node.tsx:
```diff
@@ -16,6 +16,7 @@ import { useDrag } from '../utils/useDrag';
 export interface NodeProps {
   id: string;
   theme: Theme;
+  parent?: string;
```
Couldn't nodes have more than 1 parent? I think you will need to use the graph object to get incoming edges.

amcdnl commented on docs/assets/demo.ts:
```diff
@@ -39,7 +49,7 @@ export const parentNodes: GraphNode[] = [
         id: `n-2-n-${i}`,
         label: `Node 2 > ${i} - 1 Child`,
-        parent: 'n-2',
+        parents: ['n-2'],
```
Does the user have to specify this? If so we need to remove this and use the ngraph function for getting outgoing links - 

https://github.com/anvaka/ngraph.graph/blob/main/index.js#L174

```
const links = graph. getLinks(nodeId);
const incoming = links.filter(l => l.to === nodeId);
// incoming = parents
```

jweinstein92 replied:
do we want to apply this to all nodes (allowing for all nodes with links to be collapsible and then necessitating a property allowing collapsing or not)? the original intention by manually specifying it would be to allow only specified nodes to be collapsible

jweinstein92 replied:
also, by not specifying the parents, we would then need to generating edges for these collapsible children nodes (which is currently not needed with the parents field)

jweinstein92 replied:
so yeah, seems like this is a matter of creating nodes/edges like normal. and having a property allowing collapsing/expanding of nodes with outgoing links

jweinstein92 commented:
Updated to handle parent identification internally. Also identify which nodes/edges should be hidden and instead of removing them from the graph entirely, give them a property that will hide them.

https://user-images.githubusercontent.com/6655992/187037014-cbed7027-2c18-484d-8523-4b1d0f306631.mov

jweinstein92 commented:
Updated to collapse via context menu. added stories to context menu and radial menu
https://user-images.githubusercontent.com/6655992/187047127-8ed704ec-da49-4d9c-b5cd-1d899d75c9e0.mov

amcdnl commented on src/store.ts:
```diff
@@ -57,5 +73,104 @@ export const createStore = () =>
+      }),
+    setCollapsedNodeIds: (nodeIds = []) =>
+      set(state => {
+        let collapsedNodeIds = [];
```
This needs to be moved outside of this into its own method. If you can easily add a unit test, lets do that too.

jweinstein92 replied:
Moved this into its own utility file. Was going to make a unit test for it, but the repo is currently not set up for testing as is. Would you like me to get that all set up, or just ignore it?

amcdnl commented on src/utils/store.ts:
```diff
@@ -0,0 +1,120 @@
+import { Graph } from 'ngraph.graph';
```
don't love this file name as it couples this logic that is somewhat generic to the store 'context' but its fine for now.
